client/image_list_test.go

```diff
@@ -158,3 +159,66 @@ func TestImageListApiBefore125(t *testing.T) {
 		t.Fatalf("expected 2 images, got %v", images)
 	}
 }
+
+// Checks if shared-size query parameter is set/not being set correctly
+// for /images/json.
+func TestImageListWithSharedSize(t *testing.T) {
+	// performs ImageList call and returns the sent url query
+	getImageListAndGetQuery := func(version string, options types.ImageListOptions) (query url.Values, err error) {
+		client := &Client{
+			client: newMockClient(func(req *http.Request) (*http.Response, error) {
+				query = req.URL.Query()
+
+				return &http.Response{
+					StatusCode: http.StatusOK,
+					Body:       io.NopCloser(strings.NewReader("[]")),
+				}, nil
+			}),
+			version: version,
+		}
+
+		_, err = client.ImageList(context.Background(), options)
+		return query, err
+	}
+
+	const sharedSize = "shared-size"
+
+	// Check if the shared-size query parameter is unset
+	for _, tC := range []struct {
+		name    string
+		version string
+		options types.ImageListOptions
+	}{
+		{name: "after 1.42, if not requested", version: "1.42", options: types.ImageListOptions{SharedSize: false}},
+		{name: "before 1.42, even if requested", version: "1.41", options: types.ImageListOptions{SharedSize: true}},
+	} {
+		t.Run("unset "+tC.name, func(t *testing.T) {
+			query, err := getImageListAndGetQuery(tC.version, tC.options)
+			if err != nil {
+				t.Error(err)
+			}
+
+			if query.Has(sharedSize) {
+				t.Errorf("expected %q to be unset", sharedSize)
+			}
+		})
+	}
+
+	t.Run("set after 1.42, if requested", func(t *testing.T) {
+		query, err := getImageListAndGetQuery("1.42", types.ImageListOptions{SharedSize: true})
+		if err != nil {
+			t.Error(err)
+		}
+
+		if !query.Has(sharedSize) {
+			t.Errorf("expected %q to be set", sharedSize)
+		}
+
+		actual := query.Get(sharedSize)
+		wanted := "1"
+		if actual != "1" {
+			t.Errorf("expected %q = %q, got: %q", sharedSize, wanted, actual)
+		}
```

Comment:
While it's good practice to not use test-dependencies for small packages / modules that will be used externally, it looks like we're already using `gotest.tools` in the client, so we may as well use it.

Reply:
Yeah I was tempted to use it, but noticed that the rest of tests in this file use the primitive `t.Errorf` approach, so wanted to keep this consistent.
In this case, I'll be happy to change it to use `gotest.tools` :)

Reply:
Thanks! Yes it's good to look if it's already in use; I assumed you looked at the rest of the file (👍); in this case I looked at the package as a whole; and it was already widely in use, so felt ok to use it here as well.

In cases where it's only _minimally_ used, I tend to be more careful (and consider the opposite: remove the dependency).